pkgs/build-support/wrapper-common/utils.bash

```diff
@@ -69,9 +69,9 @@ badPath() {
     # directory (including the build directory).
     test \
         "$p" != "/dev/null" -a \
-        "${p:0:${#NIX_STORE}}" != "$NIX_STORE" -a \
-        "${p:0:4}" != "/tmp" -a \
-        "${p:0:${#NIX_BUILD_TOP}}" != "$NIX_BUILD_TOP"
+        "${p#${NIX_STORE}}" = "$p" -a \
+        "${p#${TMP:-/tmp}}" = "$p" -a \
+        "${p#${NIX_BUILD_TOP}}" = "$p"
```

Comment:
This looks alright but feels harder to reason about. Compare:

* beginning of `p` is not `x`
* `p` with `x` removed from the start (if there) is still `p`

Reply:
I tried quite a few things and spent some time in the man page until I finally managed to come up with that. The really hard part was trying to come up with something that would let you default `TMP` to `/tmp`.  :grin: 

Essentially we use the _strip leading prefix of p_ expansion (`${p#<prefix to try and strip>}`} with each of our path prefixes we want the check
* if it manages to strip a leading prefix (string is altered) then we had a leading prefix match and the path is okay
* if it doesn't manage to strip a leading prefix (string isn't altered) then we didn't have a leading prefix match and the path is bad

This is quite flexible as the `<prefix to try and strip>` is itself expanded, so get to use all the power of expansion again when making our prefix to try and strip (in this case, the critical one being we can default `TMP`).